config/dev.exs

```diff
@@ -4,7 +4,7 @@ import Config
 config :lightning, Lightning.Repo,
   username: "postgres",
   password: "postgres",
-  hostname: "localhost",
+  hostname: System.get_env("POSTGRES_HOST") || "localhost",
```

Comment:
The reason was because we get this error (on Lightning `main` branch):
```
** (Mix) Could not start application lightning: Lightning.Application.start(:normal, [:timex]) returned an error: shutdown: failed to start child: Lightning.Repo
    ** (EXIT) an exception was raised:
        ** (Ecto.InvalidURLError) invalid url "postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${POSTGRES_HOST}:${POSTGRES_PORT:-5432}/${POSTGRES_DB}", host is not present. The parsed URL is: %URI{scheme: nil, userinfo: nil, host: nil, port: nil, path: "\"postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${POSTGRES_HOST}:${POSTGRES_PORT:-5432}/${POSTGRES_DB}\"", query: nil, fragment: nil}
            (ecto 3.10.3) lib/ecto/repo/supervisor.ex:92: Ecto.Repo.Supervisor.parse_url/1
            (ecto 3.10.3) lib/ecto/repo/supervisor.ex:28: Ecto.Repo.Supervisor.runtime_config/4
            (ecto 3.10.3) lib/ecto/repo/supervisor.ex:177: Ecto.Repo.Supervisor.init/1
            (stdlib 5.0.2) supervisor.erl:330: :supervisor.init/1
            (stdlib 5.0.2) gen_server.erl:962: :gen_server.init_it/2
            (stdlib 5.0.2) gen_server.erl:917: :gen_server.init_it/6
            (stdlib 5.0.2) proc_lib.erl:241: :proc_lib.init_p_do_apply/3
```
when running `iex -S mix phx.server` having the DATABASE_URL in the .env

Reply:
Got this firstly on the umbrella app but can reproduce with Lightning standalone on main branch.

Reply:
Why are there unexpanded shell vars in there? 

I tried this in Lightning:

```diff
diff --git a/config/runtime.exs b/config/runtime.exs
index ec27be1da..67dee4042 100644
--- a/config/runtime.exs
+++ b/config/runtime.exs
@@ -211,6 +211,9 @@ if log_level = System.get_env("LOG_LEVEL") do
   end
 end
 
+config :lightning, Lightning.Repo,
+  url: System.get_env("DATABASE_URL")
+
 if config_env() == :prod do
   database_url =
     System.get_env("DATABASE_URL") ||
```

And:

```sh
DATABASE_URL=postgres://postgres:postgres@localhost/lightning_test mix test
```

Works for me, and fails (correctly if I change the username or password).

I also verified using `.env.{test|dev}` and it works with expansion:

```
POSTGRES_USER="postgres"
POSTGRES_PASSWORD="postgres"
POSTGRES_HOST="localhost"
POSTGRES_DB="lightning_dev"
DATABASE_URL=postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${POSTGRES_HOST}:${POSTGRES_PORT:-5432}/${POSTGRES_DB}
```

In Lightning we use `Vapor` which I assume is doing the shell expansion for us (because personally I don't use `dotenv` etc.).

Reply:
Alright I know the issue, the loading of _all_ env vars happens with `Vapor.load!` in runtime.exs.

It does not do shell expansions (which isn't that surprising). Those env variables (the `POSTGRES_*` ones) were added _ages_ ago, and aren't really used except for filling in `DATABASE_URL`.

My vote is we remove the `POSTGRES_*` vars and put the values into the `DATABASE_URL` directly.

It's not a perfect solution, because if we loaded an env file into Docker it would do the expansion for us, or similarly if we expected everyone to use a tool like `dotenv`.

But for now, I think there is nothing lost here since we don't use those keys (or rather don't have to use this) and they are the only ones in the `.env` file that uses shell expansion.

Reply:
As the POSTGRES_* are needed. There will be another PR to restructure the use of postgres env vars.